File: client/Assets/Plugins/ProtobufTextSerializer/Scanner/Tokenizer.cs

namespace ProtobufText
{
    class Tokenizer
    {
        string _source;
       
        public Tokenizer(string src)
        {
            _source = src;
        }

        public override string ToString()
        {
            return Current.ToString();
        }

        public int Index { get; set; }

        public string Source { get { return _source; } }

        public char Current
        {
            get {
                if (EOF(0))
                    return '\0';

                return _source[Index];
            }
        }

        public int CharLeft
        {
            get { return _source.Length - Index;  }
        }

        public char Peek(int offset)
        {
            if (EOF(offset))
                return '\0';

            return _source[Index + offset ];
        }

        public int Line { get; set; }


        public void Consume( int count = 1 )
        {
            Index+= count ;
        }

        public bool EOF(int offset = 0)
        {
            return Index + offset >= _source.Length;
        }

        public void IncLine( )
        {
            Line++;
        }
    }
}

File: server/src/github.com/davyxu/golexer/tokenizer.go
package golexer

type Tokenizer struct {
	src   []rune
	index int
	lex   *Lexer
}

func (self *Tokenizer) Current() rune {

	if self.EOF() {
		return 0
	}

	return self.src[self.index]
}

func (self *Tokenizer) Index() int {
	return self.index
}

func (self *Tokenizer) Count() int {
	return len(self.src)
}

func (self *Tokenizer) Line() int {
	return self.lex.pos.Line
}

func (self *Tokenizer) Peek(offset int) rune {

	if self.index+offset >= len(self.src) {
		return 0
	}

	return self.src[self.index+offset]
}

func (self *Tokenizer) ConsumeOne() {

	self.index++
	self.lex.pos.Col++
}

func (self *Tokenizer) ConsumeMulti(count int) {

	self.index += count
	self.lex.pos.Col += count
}

func (self *Tokenizer) EOF() bool {
	return self.index >= len(self.src)
}

func (self *Tokenizer) increaseLine() {
	self.lex.pos.Line++
	self.lex.pos.Col = 1
}

func (self *Tokenizer) StringRange(begin, end int) string {

	if begin < 0 {
		begin = 0
	}

	if end > len(self.src) {
		end = len(self.src)
	}

	return string(self.src[begin:end])
}

func NewTokenizer(s string, l *Lexer) *Tokenizer {

	return &Tokenizer{
		src: []rune(s),
		lex: l,
	}
}
